docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,62 @@
+The purpose of this file is to describe the functional and key architecture of
+this project. It is spiced with some rational throughout.
+
+# Slots & Tokens
+
+This library shall provide multiple tokens for the user(s). Any given slot can
+only contain one token at any given time. In case of smartcards the slot is a
+reader whilst the token is the actual smartcard. Since this project won't
+implement a separate tool for "virtual swapping" of tokens, each token is
+associated with its own slot. Thus, the library offers a set of slots with a
+token inside each of these.
+
+In order to allow the creation of new tokens using the pkcs11 means, this
+library presents one more slot/token pair than tokens currently exist. The
+additional token is a blank token that can be initialized using the C_InitToken
+function.
+(Note: This is not implemented yet)
+
+# Objects
+
+Each token can contain an arbitrary amount of keys and other object. (Note:
+other objects are not implemented yet). The keys are TPM-based keys, i.e. they
+are only known to the TPM and also the key operations are executed inside the
+TPM. The other objects are encrypted using hybrid encryption with the outer
+key being sealed to the TPM.
+
+Note: Since the TPM's object are stored (encrypted) external to the TPM on the
+HDD, it is not possible to delete them, if your attacker model allows access to
+the HDD.
+
+# Key Authorizations
+
+For each token, two TPM seal blobs exist one for the SO and one for the User.
+The authValue for these seal blobs are SO's PIN and the User's PIN. The data
+sealed inside is an intermediate key that is used to decrypt a second authValue
+stored on disk.
```

Comment:
sealed inside is a wrapping key used to encrypt other auth values OR an auth value to a tpm key. The goal of having the tpm approach is that if someone accesses the disk meta data, brute forcing the TPM key auth value would trigger DA protections. Although If its wrapped with AES 256 or 128, you would be bruteforcing 128 or 256 bits of entropy, which is less then the TPM key authvalue itself. So you wouldn't try that, you would just keep trying the data itself. I think I am open to just doing this in software.

Reply:
Yes, I actually did not include the TPM2_EncrytDescrypt2() usage in this description. I was assuming software here... ;-)

Reply:
I think we want to keep TPM based wrapping if possible, if only for a small subset and for DA protections and rate limiting protections on that key.